pkg/kubelet/cm/dra/manager.go

```diff
@@ -0,0 +1,204 @@
+			if err != nil || client == nil {
+				return fmt.Errorf("failed to get DRA Plugin client for plugin name %s, err=%+v", driverName, err)
+			}
+
+			response, err := client.NodePrepareResource(context.Background(), resourceClaim.Namespace, resourceClaim.UID, resourceClaim.Name, resourceClaim.Status.Allocation.ResourceHandle)
+			if err != nil {
+				return fmt.Errorf("NodePrepareResource failed, claim UID: %s, claim name: %s, resource handle: %s, err: %+v", resourceClaim.UID, resourceClaim.Name, resourceClaim.Status.Allocation.ResourceHandle, err)
+			}
+			klog.V(3).InfoS("NodePrepareResource succeeded", "response", response)
+
+			annotations, err := generateCDIAnnotation(resourceClaim.UID, driverName, response.CdiDevice)
+			if err != nil {
+				return fmt.Errorf("failed to generate container annotations, err: %+v", err)
+			}
+
+			// Cache prepared resource
+			err = m.resources.add(
+				resourceClaim.Name,
+				resourceClaim.Namespace,
+				&resource{
+					driverName:  driverName,
+					claimUID:    resourceClaim.UID,
+					claimName:   resourceClaim.Name,
+					namespace:   resourceClaim.Namespace,
+					podUIDs:     sets.NewString(string(pod.UID)),
+					cdiDevice:   response.CdiDevice,
+					annotations: annotations,
+				})
+			if err != nil {
+				return fmt.Errorf("failed to cache prepared resource, claim: %s(%s), err: %+v", resourceClaim.Name, resourceClaim.UID, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// PrepareResources calls plugin NodePrepareResource from the registered DRA resource plugins.
+func (m *ManagerImpl) PrepareResources(pod *v1.Pod, container *v1.Container) (*DRAContainerInfo, error) {
+	if err := m.prepareContainerResources(pod, container); err != nil {
+		return nil, err
+	}
+	annotations := []kubecontainer.Annotation{}
+	for _, podResourceClaim := range pod.Spec.ResourceClaims {
+		claimName := resourceclaim.Name(pod, &podResourceClaim)
+		for _, claim := range container.Resources.Claims {
+			if podResourceClaim.Name != claim {
+				continue
+			}
+			resource := m.resources.get(claimName, pod.Namespace)
+			if resource == nil {
+				return nil, fmt.Errorf(fmt.Sprintf("unable to get resource for namespace: %s, claim: %s", pod.Namespace, claimName))
+			}
+			klog.V(3).InfoS("add resource annotations", "claim", claimName, "annotations", resource.annotations)
+			annotations = append(annotations, resource.annotations...)
+		}
+	}
+	return &DRAContainerInfo{Annotations: annotations}, nil
+}
+
+func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// Call NodeUnprepareResource RPC for every resource claim referenced by the pod
+	for _, podResourceClaim := range pod.Spec.ResourceClaims {
+		claimName := resourceclaim.Name(pod, &podResourceClaim)
+		resource := m.resources.get(claimName, pod.Namespace)
+		if resource == nil {
+			return fmt.Errorf("failed to get resource for namespace %s, claim %s", pod.Namespace, claimName)
+		}
```

Comment:
Is this idempotent? What if `UnprepareResources()` is called again times after successfuly unpreparing the resource? Won't this return an error in that case?

Reply:
replaced with ```continue```. Thank you for pointing it out. 
Currently it's almost impossible to call it after pod is terminated, but it's a valid point.